### ISSUE TITLE
Uptrace can't connect to Clickhouse cloud 

### ISSUE THREAD

**Issue by bcarlock-mycarrier** [uptrace/uptrace#317]
We're trying to set up an Uptrace instance as a proof of concept and connect it to a Clickhouse cloud database. We've confirmed that the cluster can connect to the database and that the credentials are correct, but all we get are the logs below until the container eventually restarts.

```
+ '[' 0 -eq 0 ]
+ /uptrace '--config=/etc/uptrace/uptrace.yml' pg wait
2024-01-23T16:50:32.418Z	info	command/pg.go:40	PostgreSQL is up and runnining
+ /uptrace '--config=/etc/uptrace/uptrace.yml' ch wait
2024-01-23T16:50:32.525Z	info	command/ch.go:35	ClickHouse is down	{"error": "EOF", "addr": "tm849a32za.us-central1.gcp.clickhouse.cloud:8443", "user": "default"}
2024-01-23T16:50:33.573Z	info	command/ch.go:35	ClickHouse is down	{"error": "EOF", "addr": "tm849a32za.us-central1.gcp.clickhouse.cloud:8443", "user": "default"}
2024-01-23T16:50:34.618Z	info	command/ch.go:35	ClickHouse is down	{"error": "EOF", "addr": "tm849a32za.us-central1.gcp.clickhouse.cloud:8443", "user": "default"}
2024-01-23T16:50:35.666Z	info	command/ch.go:35	ClickHouse is down	{"error": "EOF", "addr": "tm849a32za.us-central1.gcp.clickhouse.cloud:8443", "user": "default"}
... (continues until the container finally restarts)
```
We've thus far been unable to find any logging or other information that may be helpful in troubleshooting this issue.

**Comment by smunukutla-mycarrier** [comment]
+1

**Comment by smunukutla-mycarrier** [comment]
Config dump from Uptrace instance:

`# ./uptrace config dump
# /etc/uptrace/uptrace.yml

ch:
  dsn: ""
  addr: tm849a32za.us-central1.gcp.clickhouse.cloud:8443
  user: default
  password: <removed>
  database: <removed>
  query_settings: {}
  tls:
    ca_file: ""
    cert_file: ""
    key_file: ""
    min_version: ""
    max_version: ""
    insecure_skip_verify: true
    server_name_override: ""
  max_execution_time: 30s
pg:
  dsn: ""
  addr: <removed>
  user: <removed>
  password: <removed>
  database: <removed>
  tls:
    ca_file: ""
    cert_file: ""
    key_file: ""
    min_version: ""
    max_version: ""
    insecure_skip_verify: true
    server_name_override: ""
  conn_params: {}
projects:
  - id: 1
    name: Uptrace
    token: project1_secret_token
    pinned_attrs:
      - service_name
      - host_name
      - deployment_environment
    group_by_env: false
    group_funcs_by_service: false
    prom_compat: true
  - id: 2
    name: My project
    token: project2_secret_token
    pinned_attrs:
      - service_name
      - host_name
      - deployment_environment
    group_by_env: false
    group_funcs_by_service: false
    prom_compat: true
auth:
  users:
    - email: uptrace@localhost
      password: uptrace
      name: Anonymous
      avatar: ""
      notify_by_email: true
  cloudflare: []
  oidc: []
metrics_from_spans:
  - name: uptrace_tracing_spans
    description: Spans duration (excluding events)
    instrument: histogram
    unit: microseconds
    value: _duration / 1000
    attrs:
      - _system
      - _group_id
      - service_name
      - host_name
      - _status_code
    annotations:
      - display_name
    where: _event_name = ''
  - name: uptrace_tracing_events
    description: Events count (excluding spans)
    instrument: counter
    unit: "1"
    value: _count
    attrs:
      - _system
      - _group_id
      - _name
      - host_name
    annotations:
      - display_name
    where: _is_event = 1
ch_schema:
  compression: Default
  replicated: true
  cluster: default
  spans:
    storage_policy: default
    ttl_delete: 30 DAY
  metrics:
    storage_policy: default
    ttl_delete: 90 DAY
listen:
  http:
    addr: :14318
    tls: null
  grpc:
    addr: :14317
    tls: null
  tls: null
site:
  addr: http://localhost:14318
spans:
  buffer_size: 16000
  batch_size: 4000
metrics:
  drop_attrs:
    - telemetry.sdk.language
    - telemetry.sdk.name
    - telemetry.sdk.version
  buffer_size: 16000
  batch_size: 4000
  cum_to_delta_size: 40000
servicegraph:
  disabled: false
  store:
    size: 400000
    ttl: 5s
smtp_mailer:
  enabled: false
  host: localhost
  port: 1025
  auth_type: PLAIN
  username: mailhog
  password: mailhog
  from: uptrace@localhost
uptrace_go:
  disabled: false
  dsn: ""
  tls: null
logging:
  level: DEBUG
secret_key: <removed>
debug: true
telegram:
  bot_token: ""`

**Comment by bcarlock-mycarrier** [comment]
We got it to work by changing the clickhouse block to this:
```
ch:
  dsn: ""
  addr: tm849a32za.us-central1.gcp.clickhouse.cloud:9440
  user: uptrace
  password: 
  database: uptrace
  query_settings: {}
  tls:
    ca_file: ""
    cert_file: ""
    key_file: ""
    min_version: ""
    max_version: ""
    insecure_skip_verify: true
    server_name_override: "tm849a32za.us-central1.gcp.clickhouse.cloud"
  max_execution_time: 90s
```

**Comment by vmihailenco** [comment]
I'd imagine this should also work:

```
ch:
  addr: tm849a32za.us-central1.gcp.clickhouse.cloud:9440
  user: uptrace
  password: TODO
  database: uptrace
  tls: {}
```

**Comment by bcarlock-mycarrier** [comment]
> I'd imagine this should also work:
> 
> ```yaml
> ch:
>   addr: tm849a32za.us-central1.gcp.clickhouse.cloud:9440
>   user: uptrace
>   password: TODO
>   database: uptrace
>   tls: {}
> ```
> 
> See [docs](https://uptrace.dev/get/config.html#tls-client) for various options.

Unfortunately, it doesn't. Clickhouse cloud works over https with a port, but the native protocol doesn't route if the port is part of the URI. My guess is they're using something that uses SNI routing and whatever that is isn't able (or configured) to handle requests to the native protocol interface that contain the port as part of the address. 

Tl;dr: We had to add the server_name_override setting because the code from your snippet didn't work due to (probably) routing configuration at Clickhouse cloud.

**Comment by vmihailenco** [comment]
@bcarlock-mycarrier I am bit confused, because `insecure_skip_verify` and `server_name_override` seems to be mutual exclusive.

```
	// InsecureSkipVerify controls whether a client verifies the server's
	// certificate chain and host name. If InsecureSkipVerify is true, crypto/tls
	// accepts any certificate presented by the server and any host name in that
	// certificate. In this mode, TLS is susceptible to machine-in-the-middle
	// attacks unless custom verification is used. This should be used only for
	// testing or in combination with VerifyConnection or VerifyPeerCertificate.
	InsecureSkipVerify [bool](https://pkg.go.dev/builtin#bool)
```

Am I missing something or one of these 2 options may be omitted? I am asking because I'd like to update the docs.

**Comment by bcarlock-mycarrier** [comment]
Insecure_skip_verify can probably be disabled as long as TLS is still on, honestly we only enabled it to troubleshoot. I'll give it a shot and report back.

**Comment by bcarlock-mycarrier** [comment]
Confirmed- insecure_skip_verify was not actually doing anything. Disabling it made no difference and Uptrace is still connected to the Clickhouse cloud instance.

**Comment by mkherlakian** [comment]
Thank you @bcarlock-mycarrier! I upgraded from 1.5.x to 1.7 and the CH connection to CH cloud stopped working - your config changes fixed it!

**Comment by bcarlock-mycarrier** [comment]
> Thank you @bcarlock-mycarrier! I upgraded from 1.5.x to 1.7 and the CH connection to CH cloud stopped working - your config changes fixed it!

Glad it helped!
